Fix internal assert failure for torch.all and torch.any with requires_grad=True

IvanYashchuk opened the pull request:
This PR fixes https://github.com/pytorch/pytorch/issues/58547.
I added an OpInfo-based test that fails on master and passes with the
proposed changes.



cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry

IvanYashchuk committed:
Fix internal assert failure for torch.all and torch.any

This PR fixes https://github.com/pytorch/pytorch/issues/58547.
I added an OpInfo-based test that fails on master and passes with the
proposed changes.

albanD commented on test/test_ops.py:
```diff
@@ -665,6 +665,19 @@ def test_inplace_forward_mode_AD(self, device, dtype, op):
+
+        samples = op.sample_inputs(device, dtype, requires_grad=True)
+        sample = samples[0]
+        result = op(sample.input, *sample.args, **sample.kwargs)
```
Do we want to check that the result does not require gradients?

IvanYashchuk replied:
If the result has integer dtype, then it would have `requires_grad=False`. But if the function entry has "not_implemented" in derivatives.yml, then the result would have `requires_grad=True` and would raise an error only when backward is called.
The goal of this test is to verify that forward mode doesn't raise errors.

albanD replied:
Yes, we really should cleanup the concept of `non-differentiable` vs `not implemented`.

albanD commented on test/test_ops.py:
```diff
@@ -665,6 +665,19 @@ def test_inplace_forward_mode_AD(self, device, dtype, op):
+    # Tests that functions that do not support autograd do not fail unexpectedly
+    @_gradcheck_ops(op_db)
+    def test_nondifferentiable(self, device, dtype, op):
+        # Test only when supports_autograd=True and dtype is float64
```
=False ?

albanD reviewed (approved):
There is more cleanup we should do with not implemented/non differentiable.
Don't hesitate to ping me if you're interested in that!

albanD commented:
Importing to see if internal CI is happy with the new test.

mruberry commented on test/test_ops.py:
```diff
@@ -665,6 +665,19 @@ def test_inplace_forward_mode_AD(self, device, dtype, op):
 
         self._forward_grad_helper(device, dtype, op, self._get_safe_inplace(op.get_inplace()))
 
+    # Tests that functions that do not support autograd do not fail unexpectedly
```
Let's make this comment more specific -- do not fail unexpectedly in what context?

mruberry commented on test/test_ops.py:
```diff
@@ -665,6 +665,19 @@ def test_inplace_forward_mode_AD(self, device, dtype, op):
+    @_gradcheck_ops(op_db)
+    def test_nondifferentiable(self, device, dtype, op):
+        # Test only when supports_autograd=True and dtype is float64
+        if op.supports_autograd or dtype.is_complex:
```
I think this test can filter to ops with the desired autograd support and the desired dtypes to run the test in

mruberry replied:
I think this test can filter to ops with the desired autograd support and the desired dtypes to run the test in

mruberry commented on test/test_ops.py:
```diff
@@ -665,6 +665,19 @@ def test_inplace_forward_mode_AD(self, device, dtype, op):
+        # Test only when supports_autograd=True and dtype is float64
+        if op.supports_autograd or dtype.is_complex:
+            self.skipTest("Skipped! Autograd is supported or complex dtype.")
+        if "resize" in op.name:
```
Add the skips directly to the OpInfos along with a note as to why they're there so more skips are in one place.

mruberry commented on test/test_ops.py:
```diff
@@ -665,6 +665,19 @@ def test_inplace_forward_mode_AD(self, device, dtype, op):
+        if "resize" in op.name:
+            self.skipTest("Skipped! Cannot resize variables that require grad.")
+
+        samples = op.sample_inputs(device, dtype, requires_grad=True)
```
Note for the future: we should make sample input generation lazy so not all the sample inputs have to be generated to acquire just the first one

IvanYashchuk committed:
Filter ops for nondifferentiable test

Skip resize in OpInfos
Update test purpose description

IvanYashchuk commented:
Thank you, @mruberry, for the suggestions! `test_nondifferentiable` is now applied to filtered `op_db` that has `supports_autograd=False` and only `dtype=double` is tested for speed. I moved skipping of resize to the place where it belongs, to its OpInfo.
I updated the description of the test, we are expecting operators that do not support autograd to not raise errors in forward mode for inputs with `requires_grad=True`. `torch.all` and `torch.any` were raising internal assert failure, this test catches this situation.

mruberry reviewed (approved):
OpInfo-related changes LGTM! I'll let @albanD shepherd this PR through

albanD commented:
Awesome. I think the CI failures are relevant. I can land once they are fixed.

IvanYashchuk committed:
Merge branch 'viable/strict' into fix-58547

IvanYashchuk committed:
Revert the change to resize_'s inplace_variant

IvanYashchuk commented:
@albanD, I filed an issue about the failure https://github.com/pytorch/pytorch/issues/65945 and reverted the change that triggered the test to fail in this PR.

albanD commented:
Thanks!

albanD commented:
backward compat is broken on master, so you can ignore this one.
